File: server/server_utils.py
import uuid
from typing import List
from room import Room
from constants import MAX_PLAYER_COUNT, MIN_PLAYER_COUNT


def get_free_room(rooms: List[Room]):
    """Checks whether there is a room with a free spot or the server creates a new one"""

    is_every_room_full = True
    result_room = None

    # check if rooms exist and are empty
    for room in rooms:
        if room.is_free():
            is_every_room_full = False
            result_room = room
            break

    # create new room
    if is_every_room_full:
        room_id = str(uuid.uuid4())
        new_room = Room(id=room_id,
                        max_players=MAX_PLAYER_COUNT,
                        min_players=MIN_PLAYER_COUNT)
        rooms.append(new_room)
        result_room = new_room

    return result_room

def find_room_of_user(userId: str, rooms: List[Room]):
    for room in rooms:
        if room.is_user_present(userId):
            return room

def filter_out_userid(chat_history):
    """
    remove userId, sensitive information, not intended for client
    """

    filtered_users = [{key: val for key, val in d.items() if key != 'userid'} for d in chat_history]

    return filtered_users

File: js-server/server_utils.js
const Room = require('./room');
const { MAX_PLAYER_COUNT, MIN_PLAYER_COUNT } = require('./constants');
const crypto = require('crypto')

function get_free_room(rooms) {
    let is_every_room_full = true;
    let result_room = null;
    let room_index = -1;

    for (let i = 0; i < rooms.length; i++) {
        if (rooms[i].is_free()) {
            is_every_room_full = false;
            result_room = rooms[i];
            room_index = i;
            break;
        }
    }

    if (is_every_room_full) {
        const room_id = crypto.randomUUID();
        const new_room = new Room(room_id, MAX_PLAYER_COUNT, MIN_PLAYER_COUNT);
        rooms.push(new_room);
        room_index = rooms.length - 1;
        result_room = new_room;
    }

    return [room_index, result_room];
}

function find_room_of_user(userId, rooms) {
    for (const room of rooms) {
        if (room.is_user_present(userId)) {
            return room;
        }
    }

    return null;
}

function filter_out_userid(chat_history) {
    const filtered_users = chat_history.map(({ userid, ...rest }) => rest);
    return filtered_users;
}

function get_room_name_by_index(room_index) {
    const room_names = [
        "Spaceship",
        "Starbase",
        "Asteroid",
        "Hyperdrive",
        "Zealot",
        "Hyperion",
        "Protoss",
        "Zurk",
        "Klinğon",
        "Kaplumbağa"
    ];

    let room_name = "";

    if (room_index <= room_names.length - 1) {
        room_name = room_names[room_index] + " " + (room_index + 1);
    } else {
        room_name = String(room_index + 1);
    }

    return room_name;
}

module.exports = {
    get_free_room,
    find_room_of_user,
    filter_out_userid,
    get_room_name_by_index,
};
